contracts/exchangeIssuance/ExchangeIssuance.sol

```diff
@@ -0,0 +1,728 @@
+/*
+    Copyright 2021 Index Cooperative
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+    SPDX-License-Identifier: Apache License, Version 2.0
+*/
+pragma solidity 0.6.10;
+pragma experimental ABIEncoderV2;
+
+import { SafeMath } from "@openzeppelin/contracts/math/SafeMath.sol";
+import { Math } from "@openzeppelin/contracts/math/Math.sol";
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { ReentrancyGuard } from "@openzeppelin/contracts/utils/ReentrancyGuard.sol";
+import { IUniswapV2Router02 } from "@uniswap/v2-periphery/contracts/interfaces/IUniswapV2Router02.sol";
+import { IUniswapV2Factory } from "@uniswap/v2-core/contracts/interfaces/IUniswapV2Factory.sol";
+
+import { PreciseUnitMath } from "../lib/PreciseUnitMath.sol";
+import { UniswapV2Library } from "../../external/contracts/UniswapV2Library.sol";
+import { SushiswapV2Library } from "../../external/contracts/SushiswapV2Library.sol";
+import { ISetToken } from "../interfaces/ISetToken.sol";
+import { IBasicIssuanceModule } from "../interfaces/IBasicIssuanceModule.sol";
+import { IWETH } from "../interfaces/IWETH.sol";
+
+/**
+ * @title ExchangeIssuance
+ * @author Index Coop
+ *
+ * Contract for minting and redeeming any Set token using
+ * ETH or an ERC20 as the paying/receiving currency. All swaps are done using the best price
+ * found on Uniswap or Sushiswap.
+ *
+ */
+contract ExchangeIssuance is ReentrancyGuard {
+    
+    // TODO: use safeERC20
+    // TODO: Discuss KNC token approve function.
+    
+    using SafeMath for uint256;
+    
+    /* ============ Enums ============ */
+    
+    enum Exchange { Uniswap, Sushiswap }
+    
+    /* ============ constants ============ */
+    
+    uint256 constant private MAX_UINT96 = 2 ** 96 - 1;
+    
+    /* ============ State Variables ============ */
+
+    IUniswapV2Router02 private uniRouter;
+    address private uniFactory;
+    IUniswapV2Router02 private sushiRouter;
+    address private sushiFactory;
+
+    IBasicIssuanceModule private basicIssuanceModule;
+    address private WETH;
+
+    /* ============ Events ============ */
+
+    event ExchangeIssue(
```

Comment:
Can we add javadocs for each line item, especially to describe `_amountIn`, `_amountOut` in various function call cases?